src/libxively/gateway/xi_gw.c

```diff
@@ -0,0 +1,232 @@
+        xi_mqtt_qos_t will_qos         = 0;
+        xi_mqtt_retain_t will_retain   = 0;
+
+        ed_id_context->edge_device_context->context_data.main_context_handle = xih;
+
+        xi_connect_with_lastwill_to_impl(
+            ed_id_context->edge_device_context, "gateway no host", 0,
+            "edge device username", "edge device password", connection_timeout,
+            keepalive_timeout, session_type, will_topic, will_message, will_qos,
+            will_retain, client_callback );
+
+        /*
+         * todo_atigyi:
+         * - error handling
+         * - get corresponding credentials to application device id
+         * - call init on top layer
+         */
+    }
+    else
+    {
+        /* there is already such a context */
+    }
+
+    return state;
+
+err_handling:
+
+    XI_SAFE_FREE( ed_id_context->edge_device_id );
+    XI_SAFE_FREE( ed_id_context );
+
+    return state;
+}
+
+xi_state_t xi_disconnect_ed( xi_context_handle_t xih, const char* edge_device_id )
+{
+    printf( "%s\n", __FUNCTION__ );
+    XI_UNUSED( xih );
+
+    xi_state_t state = XI_STATE_OK;
+
+    xi_ed_id_ed_context_pair_t search_key = {.edge_device_id = ( char* )edge_device_id,
+                                             .edge_device_context = NULL};
+
+    const xi_vector_index_type_t found_index = xi_vector_find(
+        xi_globals.context_handles_vector_edge_devices,
+        XI_VEC_CONST_VALUE_PARAM( ( void* )&search_key ), _xi_ed_find_context_2_id );
+
+    if ( 0 <= found_index )
+    {
+        xi_ed_id_ed_context_pair_t* ed_id_context =
+            ( xi_ed_id_ed_context_pair_t* )xi_vector_get(
+                xi_globals.context_handles_vector_edge_devices, found_index );
+
+        state = xi_shutdown_connection_impl( ed_id_context->edge_device_context );
+    }
+
+    return state;
+}
+
+extern xi_state_t xi_remove_ed( xi_context_handle_t xih, const char* edge_device_id )
+{
+    printf( "%s\n", __FUNCTION__ );
+    XI_UNUSED( xih );
+
+    xi_ed_id_ed_context_pair_t search_key = {.edge_device_id = ( char* )edge_device_id,
+                                             .edge_device_context = NULL};
+
+    const xi_vector_index_type_t found_index = xi_vector_find(
+        xi_globals.context_handles_vector_edge_devices,
+        XI_VEC_CONST_VALUE_PARAM( ( void* )&search_key ), _xi_ed_find_context_2_id );
+
+    if ( 0 <= found_index )
+    {
+        xi_ed_id_ed_context_pair_t* ed_id_context =
+            ( xi_ed_id_ed_context_pair_t* )xi_vector_get(
+                xi_globals.context_handles_vector_edge_devices, found_index );
+
+        xi_delete_context_with_custom_layers(
+            &ed_id_context->edge_device_context, xi_gw_layer_chain,
+            XI_LAYER_CHAIN_SCHEME_LENGTH( XI_LAYER_CHAIN_GW ) );
+
+        xi_vector_del( xi_globals.context_handles_vector_edge_devices, found_index );
+
+        XI_SAFE_FREE( ed_id_context->edge_device_id );
+        XI_SAFE_FREE( ed_id_context );
+    }
```

Comment:
should this disconnect if the context hasn't already been disconnected already, or return an error if the context is currently connected?  "NO" is an acceptable answer.

Reply:
I am not here yet. I wrote this up. But the main context create/delete functions don't have this intelligent logic inside. There are the xi_create_context and xi_delete_context. And the xi_delete_context doesn't give a f@kc about the connection state of the context to be deleted.
But yes, I can make this edge device more idiot proof, but again I am not there yet. Currently I am happy seeing the whole design in action even with memory usage instabilities. So I would leave this here as it is for now if you accept, thx.

Reply:
OK. Maybe add a TODO remind or something, if you think it would help.

Or if you have to take a holiday at a critical time, it might be more obvious for JC or me to understand whats left to do.

Reply:
too many things left, but, I put there todo